### PR TITLE
Issue #13494: fixed https://checkstyle.org/checks/ & config_xxxx URLs giving 404 error

### PR DESCRIPTION
**Pull request by Zopsss**
Fixes: #13494 

added `index.xml` under `.../xdocs/checks` directory, after running `mvn clean site -Pno-validations` locally, I was able to generate `index.html` file under `.../xdocs/checks` directory. Path of that index.html file: `http://localhost:63342/checkstyle/target/site/checks/index.html`

Created `index.xml` instead of `index.html` because `index.html` gets ignored while generating the site using `mvn clean site -Pno-validations` so it does not appears in the final generated html files. Generating html files through xml files also supports `../checks.html` format.

### CODE REVIEW

**Comment by Zopsss** [comment]
Github, generate website

**Comment by Zopsss** [comment]
Github, generate website

**Comment by Zopsss** [comment]
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8217c35_2024101120/checks/index.html#

There's a small delay in redirecting, you will be able to see a blank page for a small fraction of time.

Also https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8217c35_2024101120/checks/ isn't working, normally JS picks up index.html as a default page but that is not happening in this case.

@romani @rnveach, seeking for your review and help

**Comment by Zopsss** [comment]
> JS isn't initiated when you open a url. It is only initiated after the html page is loaded.

Yeah, how can I fix this? I mean the link isn't working with `checks/` at the end, how we should tackle this?

**Comment by romani** [comment]
Links like 
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8217c35_2024101120/config_coding.html#VariableDeclarationUsageDistance

should work

there  should be page config_coding.html first of all and JS in it should do redirection after page is loaded to https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8217c35_2024101120/checks/coding/variabledeclarationusagedistance.html#VariableDeclarationUsageDistance

**Comment by Zopsss** [comment]
Github, generate website

**Comment by Zopsss** [comment]
Github, generate site

**Comment by Zopsss** [comment]
@romani I used `404.html` for redirection, I created a new `redirect.js` file to statically redirect user to correct file. Now the problem is that I'm not sure if this is going to work or not as I cannot test it locally and also if we cannot test it by generating website through GH action as it'll not match the url I provided in `redirect.js`

File: src/site/site.xml
Comment (Zopsss) [inline src/site/site.xml:49]: Adding this script in `404.xml` was giving error, I tried different ways but it was failing so I had to use `site.xml` to inject `redirect.js` in `404.xml`

Comment (Zopsss) [reply src/site/site.xml:49]: redirect.js will be injected in all generated pages, so maybe we don't need 404.xml then. But I'm not sure about this, waiting for maintainers opinion.

**Comment by Zopsss** [comment]
@romani @rnveach ping

File: src/site/site.xml
Comment (romani) [reply src/site/site.xml:49]: Js is in all existing pages. But we hit non existing. Such page use to exist.
So we need to recreate it as xml file with empty content. js in this page will trigger redirection.

In other words, page becomes real, but after page is loaded , js do redirection based on url content.

Comment (romani) [reply src/site/site.xml:49]: Here is old website https://checkstyle.sourceforge.io/version/10.10.0/
So you can walk around to see a structure.

You can build site on local by
```
git checkout checkstyle-10.10.0
mvn clean site -Pno-validations 
```

We need to restore config_xxxx.xml files

Comment (romani) [reply src/site/site.xml:49]: https://github.com/checkstyle/checkstyle/releases/tag/checkstyle-10.10.0

Such files can be found at:
https://github.com/checkstyle/checkstyle/tree/226ff1c57eaaca0bf6fcbf0674aece140c31119b/src/xdocs

Comment (Zopsss) [reply src/site/site.xml:49]: > We need to restore config_xxxx.xml files

From what I understood, are you trying to say if user comes to let's say `config_coding.html`, he should be redirected to https://checkstyle.org/checks/coding/ ?

Similarly if user comes to `config_coding#VariableDeclarationUsageDistance`, he should be redirected to https://checkstyle.org/checks/coding/variabledeclarationusagedistance.html

Im assuming, you're suggesting to restore all `config_xxx.xml` files and add redirection in them to redirect to their corresponding real page. 

But instead of restoring those files, we can add redirection in `404.xml` itself. When user will enter any `config_xxxx` url like https://checkstyle.org/config_coding.html#VariableDeclarationUsageDistance then he will hit the `404.xml` file and it will check the URL and redirect the user to its corresponding page, in this case: https://checkstyle.org/checks/coding/variabledeclarationusagedistance.html

We can add js in `404.xml` to check the URL and redirect the user, this way we will not need to restore `config_xxxx.xml` files. 

Correct me if I'm going in wrong direction.

Comment (romani) [reply src/site/site.xml:49]: >config_coding.html, he should be redirected to https://checkstyle.org/checks/coding/ ?

Yes 

>config_coding#VariableDeclarationUsageDistance, he should be redirected to https://checkstyle.org/checks/coding/variabledeclarationusagedistance.html

Yes.

>But instead of restoring those files, we can add redirection in 404.xml itself

I doubt it will work as url will be ..../404.html , you will not have original url. But if you find a way to do it, I am ok with this.

Comment (Zopsss) [reply src/site/site.xml:49]: @rnveach @romani I created a new [personal repo](https://github.com/Zopsss/checkstyle-test) in which I kept all the CS docs. I also added the [404.html](https://github.com/Zopsss/checkstyle-test/blob/master/404.html) for the redirection.

Here's the site link: https://zopsss.github.io/checkstyle-test/

Currently I have added redirection only for some specific URLs which you can check from 404.html, you can test the website with any of the following URLs and it'll redirect you to CS' URLs.

https://zopsss.github.io/checkstyle-test/checks/

https://zopsss.github.io/checkstyle-test/config_coding.html

https://zopsss.github.io/checkstyle-test/config_coding.html#VariableDeclarationUsageDistance

If we inject similar JS in our 404.xml then we'll be able to redirect users to the original URLs. The only downside of this approach I can see is, if the body of the 404.html is empty then it'll not show any kind of message ( like page not found ) to the user.

For example:

Screenshot 2024-11-28 121630

you can see in the screenshot that the URL is https://zopsss.github.io/checkstyle-test/something so it should give `404 page not found` error message, but the body of 404.html is empty so it does not display any kind of message.

We can fix this by designing our own error message, I would be happy to do that :)

Comment (Zopsss) [reply src/site/site.xml:49]: **UPDATE**

I have modified the 404.html and now it contains an error message, we can add something similar to that in our 404.xml file.

Comment (Zopsss) [reply src/site/site.xml:49]: I doubt that we can test the `404.xml` by generating website with `mvn clean site -Pno-validations` because then the URL will be different and the redirection won't happen, the only way to test is by releasing a new version.

Comment (rnveach) [reply src/site/site.xml:49]: https://docs.github.com/en/pages/getting-started-with-github-pages/creating-a-custom-404-page-for-your-github-pages-site

Comment (Zopsss) [reply src/site/site.xml:49]: Yes, this documentation explains how we can use 404.html to show custom `404 page not found` error, just the way I used it in above comment.

Comment (romani) [reply src/site/site.xml:49]: Push new code to let me see all details

Comment (Zopsss) [reply src/site/site.xml]: I have pushed the new code, I tried testing this code locally by changing the URLs to `localhost...` but it wasn't working as expected. To make it work, we will need to hit the URL from GH Pages deployment. So to check if it works, we will need to make a new release.

Is there any other way to test it?

File: src/xdocs/404.xml
Comment (Zopsss) [inline src/xdocs/404.xml:29]: used inline CSS because internal CSS was not working, not sure why.

**Comment by romani** [comment]
GitHub, generate website

File: src/site/site.xml
Comment (Zopsss) [reply src/site/site.xml]: In the new code, I tried to add the JS by using `<script>` tag but I just noticed after generating the website, the `<script>` I added is removed:

Screenshot 2024-11-30 095135

To inject JS in `404.html`, we can use this approach: https://github.com/checkstyle/checkstyle/commit/a7f0b2b8f80b1861467b223f1f0d5a11a516f33f

add the script in `site.xml` which will inject it in all the pages. But this will downgrade the performance of our website, we dont want this JS to be injected in all HTML pages, we want to inject it only in `404.html` page. Injecting this JS in all pages means that we will execute it every time a page loads, which might affect the loading time of our website.

Comment (romani) [reply src/site/site.xml]: Micro code is ok

**Comment by Zopsss** [comment]
@romani currently, the 404.html will not work. Please take a look at these comments: https://github.com/checkstyle/checkstyle/pull/15949#discussion_r1864069236 https://github.com/checkstyle/checkstyle/pull/15949#discussion_r1864078539

**Comment by romani** [comment]
Ok, push code that works when you have time.

File: src/site/site.xml
Comment (Zopsss) [reply src/site/site.xml:49]: I have pushed the micro code, but to test it, we will need to make a new release.

**Comment by romani** [comment]
GitHub, generate website

File: src/site/resources/js/redirect.js
Comment (romani) [inline src/site/resources/js/redirect.js:8]: Please make matching or currentUrl to match  without domain 
To make link https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/eac018a_2024165644/config_coding.html to work

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/site/resources/js/redirect.js
Comment (Zopsss) [reply src/site/resources/js/redirect.js]: I have pushed the new code, this new code isn't working properly locally, I have just pushed it to see if it works by any chance on github or not.

**Comment by Zopsss** [comment]
GitHub, generate website

**Comment by Zopsss** [comment]
Only https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/6dde21b_2024190849/checks.html redirection is working properly, other urls like https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/6dde21b_2024190849/config_coding.html is not working.

**Comment by Zopsss** [comment]
The main reason I can think of behind why the redirection is not working.

Suppose I opened a valid page like: `https://checkstyle.org/checks/coding/index.html`

To open this page, first my browser will try to find `index.html` inside the `coding` directory, if it finds it, it will load the HTML and JS for it. If we add any redirection on this URL then browser will be able to perform it because it was able to find and load the HTML & JS of  `index.html` file inside `coding` directory, the file actually existed. 

Now if we request for an invalid page like: `https://checkstyle.org/config_coding.html`

this HTML file does not exists, browser will try to find this file but it will fail. So no HTML or JS gets loaded and we get the 404 message.

This will not happen with GH pages deployment because we can define a 404.html page to handle our 404 errors, in this case, the `config_coding.html` URL will redirect the user to 404.html page and browser will try to load its HTML & JS. 

During loading its JS, it will check the URL and redirect the user to the original URL. 

Note that, when user gets 404 error, Only the custom error message is shown to user, his URL does not changes to `404.html`,  the URL would still be `config_coding.html` and we will be able to perform the redirection.

**Comment by romani** [comment]
It is more beneficial to rely on basic technology, that will work for all hosting (we have sourceforge.io) that it still preferred website by Google search engine.

Relying on 404 feature of GitHub is sweet but we need more reliable approach.

Let's restore config_xxxx pages with empty content and doing redirection to new pages.

**Comment by Zopsss** [comment]
@romani I have added some config_xxxx.xml files, they were working locally, hope they will work here too.

I haven't tested the https://checkstyle.org/checks/ locally, I'll do that soon. In meantime, please review other new config_xxxx.xml files' redirections.

File: src/site/site.xml
Comment (Zopsss) [inline src/site/site.xml:49]: For the config_xxxx.xml approach, we will still require this extra JS file as we cannot write JS inside `<script>` tags in xml files. And we are importing this JS file in all the generated HTML files, I'm not sure how we can import this JS file into only required html files, i.e config_coding.html, config_header.html, etc.

**Comment by Zopsss** [comment]
Github, generate website

**Comment by Zopsss** [comment]
GitHub, generate website

**Comment by Zopsss** [comment]
Please test the redirections with these URLs:

- https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/6ee7f4a_2024202316/config_coding.html
- https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/6ee7f4a_2024202316/config_coding.html#VariableDeclarationUsageDistance
- https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/6ee7f4a_2024202316/config_header.html
- https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/6ee7f4a_2024202316/config_header.html#RegexpHeader

The redirection takes few milliseconds so you will see a blank page for that moment.

File: src/xdocs/config_coding.xml
Comment (romani) [inline src/xdocs/config_coding.xml:9]: Can we do title  as "Redirection to xxxxx" and in our ja code skip calculations based on page title 
https://stackoverflow.com/a/1057065/1015848

**Review by romani — CHANGES_REQUESTED** [review]
Prove of concept works , good.
Small blink of page is fine, it is a way better than 404 page. I hope we can improve later on it, to be better.

Items:

File: src/site/resources/js/redirect.js
Comment (romani) [inline src/site/resources/js/redirect.js:8]: endsWith("/checks/")

File: src/site/resources/js/redirect.js
Comment (romani) [inline src/site/resources/js/redirect.js:11]: Can we have map  with routing content "from-to"

"config_coding.html#VariableDeclarationUsageDistance" to "coding/variabledeclarationusagedistance.html#VariableDeclarationUsageDistance"

And we simply iterate over map entries and in case of match by endsWith choose destination.

so

File: src/site/resources/js/redirect.js
Comment (romani) [inline src/site/resources/js/redirect.js:22]: We can not hardcode `https://checkstyle.org/`, please take it from URL object

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/site/resources/js/redirect.js
Comment (Zopsss) [reply src/site/resources/js/redirect.js]: I have created a map which contains page title as key and its corresponding redirecting function as value, then used page title to get the redirect function from map and called it:

https://github.com/checkstyle/checkstyle/blob/c9a74285db34d2963947abf6effa07f610334335/src/site/resources/js/redirect.js#L12-L18


> And we simply iterate over map entries and in case of match by endsWith choose destination.

To use `endsWith` we will require to use if/else statement but I have used switch statements, using if/else has 2 downsides:

1. The code will become uglier, hard to read considering there will be a long list of if/else-if ladder
2. Switch statements are more efficient than if/else statements plus they make code look more clean.

So I think using switch statements is a good option for our usecase.

Comment (Zopsss) [reply src/site/resources/js/redirect.js]: The code is looking more cleaner and maintainable with current implementation.

File: src/site/resources/js/redirect.js
Comment (Zopsss) [reply src/site/resources/js/redirect.js]: done. 

Had to create `index.xml` under `checks` directory to make it work. 

https://github.com/checkstyle/checkstyle/blob/c9a74285db34d2963947abf6effa07f610334335/src/site/resources/js/redirect.js#L3-L6

File: src/xdocs/config_coding.xml
Comment (Zopsss) [reply src/xdocs/config_coding.xml:9]: I have changed the page titles to `config_coding` & `config_header`, these title helps us in redirection. For better explanation, pls refer https://github.com/checkstyle/checkstyle/pull/15949#discussion_r1867078104

**Comment by Zopsss** [comment]
> It is more beneficial to rely on basic technology, that will work for all hosting (we have sourceforge.io) that it still preferred website by Google search engine.

I don't think our current redirections will also work for sourceforge site, i.e for urls like https://checkstyle.sourceforge.io/config_coding.html or https://checkstyle.sourceforge.io/config_coding.html#VariableDeclarationUsageDistance

but I'm not sure cuz now I have implemented dynamic URLs:

https://github.com/checkstyle/checkstyle/blob/027ff10829b98f9e3abbc997324d718e9b06cd2d/src/site/resources/js/redirect.js#L22-L29

File: src/site/resources/js/redirect.js
Comment (Zopsss) [reply src/site/resources/js/redirect.js]: done 

https://github.com/checkstyle/checkstyle/blob/027ff10829b98f9e3abbc997324d718e9b06cd2d/src/site/resources/js/redirect.js#L22-L29

Comment (Zopsss) [reply src/site/resources/js/redirect.js]: thanks for the suggestion! I didn't knew such dynamic routing works with `window.location.replace()` too.

**Comment by romani** [comment]
@Zopsss , please look at CI and keep fixing it, we are finishing PR.

**Comment by Zopsss** [comment]
GitHub, generate website

File: src/xdocs/config_coding.xml
Comment (romani) [reply src/xdocs/config_coding.xml:9]: We need some word marker that will help us to return quickly from redirection function, it  it executed in active page.

Comment (Zopsss) [reply src/xdocs/config_coding.xml:9]: Can you elaborate it more? Im not able to understand it.

File: src/site/resources/js/redirect.js
Comment (Zopsss) [reply src/site/resources/js/redirect.js]: https://github.com/checkstyle/checkstyle/blob/c8397ba9975a92001c25d9cae4232b5cc9e07685/src/site/resources/js/redirect.js#L3-L6

made the redirection relative.

File: src/xdocs/config_coding.xml
Comment (romani) [reply src/xdocs/config_coding.xml:9]: Title should be `Redirection to coding.html`
`If (!title.startWith("Redirection to ") ) return;`

**Comment by Zopsss** [comment]
Please remove the trailing `#` from the generated site URLs to to make redirection working.

File: src/xdocs/config_coding.xml
Comment (Zopsss) [reply src/xdocs/config_coding.xml:9]: done.

File: src/site/resources/js/redirect.js
Comment (Zopsss) [reply src/site/resources/js/redirect.js:2]: done.

**Comment by Zopsss** [comment]
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c8397ba_2024065122/config_coding.html redirects to https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c8397ba_2024065122/checks/coding/

but that page is not working. Not sure why it is happening, cuz locally everything was working fine. The https://checkstyle.org/checks/coding/ URL of main site also works fine.

and Same problem with `/config_header.html` redirection....

File: src/site/resources/js/redirect.js
Comment (Zopsss) [reply src/site/resources/js/redirect.js:2]: I have checked that if the URL is for `/checks/`, or is URL for redirection page? If both are false then we will automatically exit from this JS.

**Comment by Zopsss** [comment]
It seems like when we generate website using `GitHub, generate website`, the following of URLs https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c8397ba_2024065122/checks/coding/ requires `index.html` at the end. 

In other words, only the  following URL will work https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c8397ba_2024065122/checks/coding/index.html

But we don't require adding trailing `index.html` while generating the site locally as well as when the site is deployed.

**Comment by romani** [comment]
Let's not let this nuance of index.html handling in AWS s3 web site to block this update.
It is ok to not be able to test this url. I can deploy it GitHub.io or source forge right before merge of PR to make sure it works. One time test at the end is ok.
Let's focus on fixing all other links.

File: src/site/resources/js/redirect.js
Comment (romani) [inline src/site/resources/js/redirect.js:27]: Interesting idea:
If we do update like this for all links, it will be a lot, so js file will be with thousands lines, might not be quick to download 
:).

What if we just by code convert "#VariableDeclarationUsageDistance"
To "variabledeclarationusagedistance.html".
So at least js size will be small, so quick to download.

- [ ]

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/site/resources/js/redirect.js
Comment (Zopsss) [reply src/site/resources/js/redirect.js]: Done. thanks a lot for the suggestion. Now our JS has reduced by a lot. I deleted the `redirect.js` and moved the code to `checkstyle.js` because our JS became so small.

**Comment by Zopsss** [comment]
Proof that site redirections works locally:


https://github.com/user-attachments/assets/50fd9ffa-8e92-405f-9f27-309b3abdd52f


@romani @rnveach please do final review, if approved then I will add the remaining config_xxxx.xml files.

**Comment by romani** [comment]
Looks good, please do for rest of pages.

**Comment by Zopsss** [comment]
Github, generate website

**Comment by romani** [comment]
https://romani.github.io/site/checks

https://romani.github.io/site/config_javadoc.html#JavadocType
https://romani.github.io/site/config_coding.html#VariableDeclarationUsageDistance

**Comment by romani** [comment]
commands to redeploy to SF:
```
# in github.io repo with already generated website
tar cfz checkstyle.github.io-redirect.tar.gz site
scp -i ~/.ssh/private_sourceforge_key checkstyle.github.io-redirect.tar.gz   "$SF_USER"@web.sourceforge.net:/home/project-web/checkstyle
ssh -t -i ~/.ssh/private_sourceforge_key "$SF_USER"@shell.sourceforge.net create


cd /home/project-web/checkstyle
cd htdocs-version/
mkdir htdocs-10.20.3-SNAPSHOT
cd htdocs-10.20.3-SNAPSHOT
mv ../../checkstyle.github.io-redirect.tar.gz .
tar -xzvf checkstyle.github.io-redirect.tar.gz 
mv site/* .
cd /home/project-web/checkstyle
ln -f -s $(pwd)/htdocs-version/htdocs-10.20.3-SNAPSHOT $(pwd)/htdocs/version/10.20.3-SNAPSHOT
```

**Review by romani — APPROVED** [review]
ok to merge.

@Zopsss , thanks a lot for very good healing of our website!!!!
All links that are created by people in web for ~!0 years are back to live.

**Comment by romani** [comment]
@nrmancuso , please review.

**Comment by Zopsss** [comment]
Glad to help the community :)

**Review by nrmancuso — APPROVED** [review]
Good stuff @Zopsss

**Pull request merged**
